### ISSUE TITLE
Strange CUDA Illegal Memory Access (700)

### ISSUE THREAD

**Issue by achalagarwal** [gunrock/gunrock#791]
**Context:**
Label Propagation benefits from unique frontier elements (so as to prevent redundant computations from a snowballing effect)
So, we need to create an Array1D to check if an element has already been added to the frontier and then prevent duplicate entries using an atomic op.

```C++
   bool already_added = atomicMax(added + dest, 1) == 1; 
   return !already_added;
```

**Summary:**

In the filter + advance fused kernel, filtering in the advance_op results in the error but filtering in the filter op works fine.

**Reproducibility:**
Always

**To Reproduce:**
https://github.com/achalagarwal/gunrock/commit/3fe2ecc1f5c9768bfe409b7c1fa41ffcb0841f30

For example:
1. Clone gunrock
2. Switch to the debug_lp branch and commit 3fe2ecc1f5c9768bfe409b7c1fa41ffcb0841f30

Make sure to:
$ git submodule init
$ git submodule update
etc.
etc.

**Expected Behavior:**
[../../gunrock/app/lp/lp_enactor.cuh, 256 @ gpu 0] cuda%%%Synchronize failed. (CUDA error 700: an illegal memory access was encountered)
[../../gunrock/app/enactor_base.cuh, 519 @ gpu 0] error encountered (CUDA error 700: an illegal memory access was encountered)
[../../gunrock/app/lp/lp_enactor.cuh, 623 @ gpu 0] error encountered (CUDA error 700: an illegal memory access was encountered)
[../../gunrock/app/lp/lp_app.cu, 100 @ gpu 0] error encountered (CUDA error 700: an illegal memory access was encountered)
... 

**CUDA Versions:**
10, 10.2, 11

**Comment by jowens** [comment]
I can't comment on how this is implemented, but I'm a little nervous about an array data structure that enforces strict uniqueness. Does our array data structure actually implement this? (I don't know.) This seems expensive. Sort-then-uniquify?

**Comment by achalagarwal** [comment]
No, I don't think there is provision for uniqueness in our Array data structure. That's why I was using another mask array with atomicMax to maintain uniqueness. 

That's not the issue though, the issue is that the same lines of code work in filter but not in advance. Perhaps @neoblizz might know as it could be related to how the fused kernel (filter + advanced) works.

**Comment by neoblizz** [comment]
@achalagarwal have you tried printing `dest`? In filter, I believe `dest` is really the input nodes coming in since all nodes are in a sense "destination". I think the naming is confusing (fixing this in the refactor). And in advance, `dest` implies the neighbors. So, `added + dest` in filter may mean `added + src node` and in advance it may mean `added + neighbor node`.

That'll be my first guess.

**Comment by achalagarwal** [comment]
> In filter, I believe `dest` is really the input nodes coming in since all nodes are in a sense "destination". I think the naming is confusing (fixing this in the refactor). And in advance, `dest` implies the neighbors. So, `added + dest` in filter may mean `added + src node` and in advance it may mean `added + neighbor node`.

Aah interesting! I will verify this, thanks!

**Comment by achalagarwal** [comment]
For me, the fused kernel itself doesn't work. I verified this crudely using print statements and checking the changes. Basically the filter op doesn't print/function when fused but when used separately it does the work


It's not critical for me as I am launching the filter op separately in my lp implementation. But I might come back to why it wasn't working in the fused kernel later.
